src/TRoomDB.cpp

```diff
@@ -551,43 +580,346 @@ QList<int> TRoomDB::getAreaIDList()
+            itRenumberedRoomId.next();
+            unsigned int newRoomId = 0;
+            do {
+                ; // Noop - needed increment is done in test condition!
+            } while( validUsedRoomIds.contains( ++newRoomId ) );
+
+            itRenumberedRoomId.setValue( newRoomId ); // Update the QHash
+            validUsedRoomIds.insert( newRoomId );
+            infoMsg.append( QStringLiteral( "%1 ==> %2" )
+                                .arg( itRenumberedRoomId.key() )
+                                .arg( itRenumberedRoomId.value() ) );
+
+        }
+        mpMap->postMessage( infoMsg );
+
+        QSet<TRoom *> holdingSet;
+        { // Block this code to limit scope of iterator
+            QMutableHashIterator<int, TRoom* > itRoom( rooms );
+            // Although this is "mutable" we can only changed the VALUE not the KEY
+            // Also we cannot directly change the collection being iterated over -
+            // though we can REMOVE items via the iterator - so pull the
+            // affected TRoom instances out of the Hash, renumber them and then
+            // hold onto them before reinserting them after we have finished
+            // this iteration
+            while( itRoom.hasNext() ) {
+                itRoom.next();
+                TRoom * pR = itRoom.value();
+                if( roomRemapping.contains( itRoom.key() ) ) {
+                    pR->userData.insert( QStringLiteral( "audit.remapped_id" ), QString::number( itRoom.key() ) );
+                    pR->setId( roomRemapping.value( itRoom.key() ) );
+                    itRoom.remove();
+                    holdingSet.insert( pR );
+                }
+            }
+        }
+
+        // Now stuff the modified values back in under the new key values
+        QSetIterator<TRoom *> itModifiedRoom( holdingSet );
+        while( itModifiedRoom.hasNext() ) {
+            TRoom * pR = itModifiedRoom.next();
+            int newRoomId = pR->getId();
+            rooms.insert( newRoomId, pR );
+        }
+    }
+    else {
+        QString infoMsg = tr( "[ INFO ]  - Room Id numbering is satisfactory." );
+        mpMap->postMessage( infoMsg );
+    }
+    // END OF TASK 1
+
+    // START OF TASK 6,7 & 9
+    { // Block the following code to limit the scope of the itRoom iterator
+        QMutableHashIterator<int, TRoom *> itRoom( rooms );
+        while( itRoom.hasNext() ) {
+            itRoom.next();
+            TRoom * pR = itRoom.value();
+
+            // Purges any duplicates that a QList structure DOES permit, but a QSet does NOT:
+            // Exit stubs:
+            unsigned int _listCount = pR->exitStubs.count();
+            QSet<int> _set = pR->exitStubs.toSet();
+            if( _set.count() < _listCount ) {
+                QString infoMsg = tr( "[ INFO ]  - Duplicate exit stub identifiers found in room Id: %1, this is an\n"
+                                                  "anomoly but has been cleaned up easily." )
```

Comment:
An exit lock is a boolean status, so what does it mean by a duplicate exit stub identifier?